[Feat] 챌린지 상단 헤더 정보 조회 API 구현

yangjiae12 opened the pull request:
## #️⃣ 연관된 이슈

> 관련된 이슈 번호를 적어주세요.\
> Close #74 

## ✨ 작업 내용 (Summary)

> 이번 PR에서 작업한 내용을 간략히 설명해주세요. (이미지 첨부 가능)

챌린지 상세 페이지의 상단 정보 조회 API를 구현한 작업입니다. 클라이언트의 연산 부담을 줄이고 데이터 일관성을 보장하기 위해, 서버에서 다양한 상태 값을 계산 및 조합하여 내려주도록 설계했습니다.

### 1. 주요 비즈니스 로직 및 계산 처리
단순 DB 조회 외에 서비스 계층에서 수행하는 연산 로직은 다음과 같습니다.

#### 날짜 및 기간 계산
- 시작/종료일: 챌린지 자체의 기간이 아닌 현재 진행 중인 라운드(`CurrentRound`)의 `startDate`와 `endDate`를 추출하여 반환합니다.
- 남은 일수(`remainDays`)
  - `ChronoUnit.DAYS`를 사용하여 오늘과 종료일 사이의 일수를 계산합니다.
  - 종료된 챌린지의 경우 음수가 나오지 않도록 최솟값을 0으로 보장했습니다.

#### 유저 상태 판별 (`User Status`)
  - 참여 여부(`isParticipant`): `UserChallenge` 테이블을 조회하여 판단합니다.
  - 좋아요 여부 (`isLiked`): `ChallengeLike` 테이블을 조회하여 Boolean 값으로 매핑했습니다.
  - 관찰자 모드 (`isObserverMode`): `Entity`의 `isViewerMode` 필드를 DTO의 `isObserverMode`로 매핑했습니다.

#### 금일 인증 여부 체크 (`isCertifiedToday`)
- 참여자인 경우에 단순히 인증 글 존재 여부만 보는 것이 아니라 아래 조건들을 보고 체크합니다.
1. 시간 범위: 챌린지에 설정된 **인증 시작 시간(`verifyStartTime`) ~ 종료 시간(`verifyEndTime`)**을 오늘 날짜에 적용하여 범위를 산정합니다.
2. 상태 체크: 임시저장(`TEMPORARY`)은 제외하고, 완료됨(`COMPLETED`) 상태인 인증 글만 카운트합니다.

#### 방장 정보 조회 (`Owner Info`)
- `UserChallenge` 테이블에서 Role이 `OWNER`인 데이터를 조회합니다.
- 성능 최적화: 유저 정보까지 한 번에 가져오기 위해 JOIN FETCH를 적용한 커스텀 쿼리(`findOwnerByChallengeId`)를 사용하여 N+1 문제를 방지했습니다.

### 2. 하단 버튼 상태 결정 (resolveButtonStatus)
- 위에서 계산된 데이터(참여 여부, 요일, 시간, 인증 여부 등)를 종합하여 최종 버튼 UI 상태를 도출합니다. (아래 표 참조)

우선순위 | 유저 상태 | 상세 조건 | 결과 (Enum) | UI 노출
-- | -- | -- | -- | --
1 | - | 챌린지 종료 (FINISHED) | DISABLED | 비활성화 (종료됨)
2 | 참여자 | 라운드 시작 전 (오늘 < 시작일) | CERTIFIED | D-Day 표시
3 | 참여자 | 인증 요일 아님 OR 시간대 아님 | CERTIFIED | D-Day 표시
4 | 참여자 | 인증 시간 내 + 이미 인증함 | CERTIFIED | D-Day 표시
5 | 참여자 | 인증 시간 내 + 미인증 | CERTIFY_AVAILABLE | (Active) 인증하기
6 | 미참여자 | 현재원 >= 정원 (만석) | WAITLIST | 빈자리 알림
7 | 미참여자 | 모집중 (UPCOMING/RECRUITING) | JOIN | 참가하기

---

## ✅ 변경 사항 체크리스트

> 다음 항목들을 확인하고 체크해주세요.

- [ ] 코드에 영향이 있는 모든 부분에 대한 테스트를 작성하고 실행했나요?
- [ ] 문서를 작성하거나 수정했나요? (필요한 경우)
- [ ] 중요한 변경 사항이 팀에 공유되었나요?

---

## 🧪 테스트 결과

> 코드 변경에 대해 테스트를 수행한 결과를 요약해주세요.

* **테스트 환경:** 로컬
* **테스트 방법:** JUnit5 + Modkito (단위 테스트), Swagger
* **결과 요약:** 단위 테스트의 핵심 시나리오 Pass, Swagger 통해 작동 확인

| 테스트 케이스 (Method Name) | 조건 (참여/시간/인증) | 결과 (Button Status)
-- | -- | --
participant_upcoming_beforeStart_returns_CERTIFIED | 참여자 + 시작 전(UPCOMING) | CERTIFIED (D-Day)
participant_notVerificationDay_returns_CERTIFIED | 참여자 + 인증 요일 아님 | CERTIFIED (D-Day)
participant_notVerificationTime_returns_CERTIFIED | 참여자 + 인증 시간 밖 | CERTIFIED (D-Day)
participant_readyToVerify_returns_AVAILABLE | 참여자 + 인증 시간 내 + 미인증 | CERTIFY_AVAILABLE (인증하기)
participant_alreadyVerified_returns_CERTIFIED | 참여자 + 인증 시간 내 + 인증완료 | CERTIFIED (완료)
guest_recruiting_hasSpace_returns_JOIN | 미참여자 + 모집중 + 자리 있음 | JOIN (참가하기)
guest_recruiting_full_returns_WAITLIST | 미참여자 + 모집중 + 만석 | WAITLIST (대기신청)
challenge_finished_returns_DISABLED | (공통) 챌린지 종료됨 | DISABLED (비활성)

---

## 📸 스크린샷

> 관련된 스크린샷 또는 GIF가 있다면 여기에 첨부해주세요.

| 기능 | 사진 |
|---|---------|
| Swagger를 통한 테스트 | <img width="1685" height="744" alt="image" src="https://github.com/user-attachments/assets/344e6ba1-fc13-4cd2-a45a-d46d232c8f3f" /> |
| 단위 테스트 | <img width="616" height="268" alt="image" src="https://github.com/user-attachments/assets/ccd7503d-62a1-4c49-a182-24f094977603" /> |

---

## 💬 리뷰 요구사항

> 리뷰어가 특별히 봐주었으면 하는 부분이 있다면 작성해주세요.
- D-Day 계산 로직
- 인증 시간 체크 로직
- Entity 필드 추가에 따른 스크립트 작성 (기존 데이터는 없음)

---

## 📎 참고 자료

> 관련 문서, 레퍼런스 링크 등이 있다면 여기에 첨부해주세요.


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **새 기능**
  * 챌린지 헤더 정보가 강화되어 소유자 정보, 참가자 수, 시작/종료일·남은 일수, 사용자별 참여/관찰/좋아요 상태 및 상황에 따른 동작 버튼 상태를 제공합니다.
  * 인증(Verification) 포스트가 텍스트/사진 타입을 구분하고 제목·본문·이미지·링크·질문 여부 및 임시/완료 상태를 저장하도록 확장되었습니다.
* **테스트**
  * 다양한 시나리오의 헤더 정보 및 버튼 상태 동작을 검증하는 단위 테스트가 추가되었습니다.
* **데이터베이스**
  * 인증 관련 필드(타입·제목·본문·이미지·링크·질문 여부·상태)를 추가하는 마이그레이션이 포함되었습니다.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

yangjiae12 committed:
feat: 인증(Verification) 엔티티 필드 확장 및 DB 마이그레이션 스크립트 추가 #74

- Verification 엔티티에 상세 정보 필드(type, title, content, photoUrl 등) 추가
- RoundRecord와의 연관관계 재설정 및 기존 UserChallenge 연관관계 제거
- 인증 유형(VerificationPostType) 및 상태(VerificationStatus) Enum 정의
- Flyway V2.4 스크립트(V2.4__add_verification_details.sql) 추가하여 DB 스키마 반영

yangjiae12 committed:
feat: 챌린지 상세 상단 정보 조회를 위한 Repository 쿼리 및 DTO 구현 #74

- ChallengeResponseDto 내부에 상단 정보용 HeaderInfoDto 및 HostDto 클래스 추가
- 하단 버튼 상태 제어를 위한 ActionButtonStatus Enum 정의
- ChallengeRepository: N+1 문제 해결을 위한 findByIdWithDays(Fetch Join) 쿼리 추가
- UserChallengeRepository: 참여 여부 및 방장 정보 조회를 위한 메서드 추가
- VerificationRepository: 오늘 날짜 기준 완료된 인증 내역 조회 쿼리 구현

yangjiae12 committed:
feat: 챌린지 상세 상단 정보 조회 비즈니스 로직 구현 #74

- ChallengeService 인터페이스에 getChallengeHeaderInfo 메서드 정의
- ChallengeServiceImpl 구현:
  - 로그인 유저의 참여 여부, 좋아요 여부, 오늘 인증 여부 복합 확인 로직 구현
  - 요일, 시간대, 인증 상태를 종합하여 ActionButtonStatus를 결정하는 핵심 로직(resolveButtonStatus) 구현
- ChallengeConverter: Entity 데이터를 HeaderInfoDto로 변환하는 매핑 로직 추가

yangjiae12 committed:
feat: 챌린지 상세 상단 정보 조회 API 연결 #74

yangjiae12 committed:
test: 챌린지 버튼 상태 결정 로직에 대한 단위 테스트 작성 #74

- ChallengeServiceImpl의 getChallengeHeaderInfo 메서드 단위 테스트(Unit Test) 작성
- Mockito를 활용하여 Repository 의존성을 모킹
- 참여자/미참여자, 인증 시간 내/외, 만석 여부 등 5가지 핵심 시나리오에 대한 버튼 상태(ActionButtonStatus) 검증

yangjiae12 committed:
fix: 챌린지 시작 전(UPCOMING) 인증 버튼이 활성화되는 버그 수정 #74

- 참여자가 챌린지 시작일 이전에 상세 페이지에 진입할 경우, 인증 시간대가 맞더라도 '인증하기' 버튼이 아닌 'D-Day' 상태(CERTIFIED)를 반환하도록 예외 로직 추가

yangjiae12 committed:
test: 챌린지 버튼 상태 결정 단위 테스트 안정성 개선 및 NPE 수정 #74

- 종료된 챌린지 테스트 시 verifyTime 설정 누락으로 인한 NullPointerException 수정
- Mockito의 Strictness 정책으로 인한 UnnecessaryStubbingException 해결을 위해 공통 헬퍼 메서드에 lenient() 적용
- 시작일 이전(UPCOMING) 케이스에 대한 검증 시나리오 보완

yangjiae12 committed:
refactor: 챌린지 요일 확인 로직 내 불필요한 방어 코드 제거 #74

- 챌린지 생성 시 @NotEmpty 어노테이션을 통해 요일 정보의 존재가 보장
- 서비스 로직 내 '요일이 없으면 매일 인증'으로 처리하던 방어 코드를 제거하여 데이터 무결성 정책 통일

yangjiae12 committed:
Merge branch 'develop' into feat/74-challenge-info

yangjiae12 committed:
chore: DB 마이그레이션 스크립트 버전 수정 #74

yangjiae12 committed:
chore: Verification 엔티티 주석 수정 #74

yc3697 commented on src/main/java/com/hrr/backend/domain/challenge/service/ChallengeServiceImpl.java:
```diff
@@ -531,9 +581,144 @@ private void validateJoinRequest(Challenge challenge, User user, String inputPas
+        );
+    }
+
+    /**
+     * 하단 버튼의 상태(ActionButtonStatus)를 결정하는 핵심 로직
+     * (기획 따라 변경 가능)
+     */
+    private ActionButtonStatus resolveButtonStatus(Challenge challenge, boolean isParticipant, boolean isCertifiedToday) {
```
프론트에서 고려할 게 너무 많겠다 싶었는데 아주 깔끔하게 잘 정리된 것 같네요 ㅎㅎ 좋습니다!

yc3697 commented on src/main/java/com/hrr/backend/domain/user/repository/UserChallengeRepository.java:
```diff
@@ -4,9 +4,28 @@
+        FROM UserChallenge uc 
+        JOIN FETCH uc.user 
+        WHERE uc.challenge.id = :challengeId 
+        AND uc.role = 'OWNER'
```
여기 문자열로 돼있는 부분만 UserChallengeRole.OWNER였나 그걸로 바꿔주시면 좋을 것 같아요

yangjiae12 replied:
파라미터로 `UserChallengeRole` 받도록 수정했습니다~!

yc3697 reviewed (approved):
꽤 복잡한 로직이었을 텐데 꼼꼼하게 잘해주셨네요. 수고하셨습니다!!

yangjiae12 committed:
refactor: 방장 조회 쿼리 하드코딩 제거 및 Enum 파라미터 적용 #74

- 기존 'OWNER' 문자열 직접 명시 방식을 제거
- 메서드 파라미터로 Role Enum을 받도록 수정하여 타입 안정성 확보
- Service 계층 호출부 및 테스트 코드 수정
